(PUP-9455) Manage agent keys and certs

joshcooper opened the pull request:
Create CertProvider to manage CA certs, CRLs, CSRs, private keys, client certs.

* PEM files are read/written in UTF-8 due to comments
* Existing behavior for names is preserved (case-insensitive and restricted character set to protect against directory traversal)
* Permissions are not yet enforced (see PUP-9463)
* Encrypted private keys are not supported (existing support for them is dubious and will be addressed later, see PUP-9466)

puppetcla commented:
CLA signed by all contributors.

melissa commented on spec/unit/x509/cert_provider_spec.rb:
```diff
@@ -0,0 +1,305 @@
+
+      it 'returns an array of certificates' do
+        certs = create_provider(capath: my_fixture('ca.pem')).load_cacerts
+        expect(certs.count).to eq(1)
```
would it be helpful to check that it's an array object rather than checking the count? Or do you think this is sufficient to verify we've been given an array?

joshcooper replied:
I updated this to use `contain_exactly` and `an_object_having_attributes`.

melissa commented on spec/unit/x509/cert_provider_spec.rb:
```diff
@@ -0,0 +1,305 @@
+        expect(File.read(ca_path)).to match(/\A-----BEGIN CERTIFICATE-----.*?-----END CERTIFICATE-----\Z/m)
+      end
+
+      it 'sets mode to 644'
```
This looks like an unfinished test

joshcooper replied:
Ah this is a pending test to be fixed in PUP-9463. (I filed the ticket after this was submitted). I'll update the test description.

melissa commented on spec/unit/x509/cert_provider_spec.rb:
```diff
@@ -0,0 +1,305 @@
+      end
+
+      it 'sets mode to 644'
+    end
```
Also, probably a good idea to define what happens if it fails to save a cacert

melissa replied:
And what happens if we have different content then an existing cert with the same name? Is it overwritten?

joshcooper replied:
I updated the CertProvider and tests to handle cases where we fail to save or load a cacert, private key, etc

melissa commented on spec/unit/x509/cert_provider_spec.rb:
```diff
@@ -0,0 +1,305 @@
+        expect(File.read(crl_path)).to match(/\A-----BEGIN X509 CRL-----.*?-----END X509 CRL-----\Z/m)
+      end
+
+      it 'sets mode to 644'
```
also unfinished test

melissa commented on spec/unit/x509/cert_provider_spec.rb:
```diff
@@ -0,0 +1,305 @@
+    context 'private keys' do
+      let(:provider) { create_provider(privatekeydir: my_fixture_dir) }
+
+      it 'returns nil if it does not exist' do
```
will it return nil or error out if the file is not readable?

joshcooper replied:
All of the CertProvider methods will raise `Puppet::Error` if there's an unexpected error like that. I want to treat the file doesn't exist vs it's unreadable as different things so that the caller doesn't have to test if the file exists before calling the provider. If we do that, then it's possible for the file to be deleted between the time we test and when we try to read it. So it's better to just try to read it, and gracefully handle the `ENOENT` exception.

melissa commented on spec/unit/x509/cert_provider_spec.rb:
```diff
@@ -0,0 +1,305 @@
+        expect(File.read(path)).to match(/\A-----BEGIN RSA PRIVATE KEY-----.*?-----END RSA PRIVATE KEY-----\Z/m)
+      end
+
+      it 'sets mode to 640'
```
unfinished test?

melissa commented on spec/unit/x509/cert_provider_spec.rb:
```diff
@@ -0,0 +1,305 @@
+        expect(File.read(path)).to match(/\A-----BEGIN CERTIFICATE-----.*?-----END CERTIFICATE-----\Z/m)
+      end
+
+      it 'sets mode to 644'
```
unfinished test?

melissa commented on spec/unit/x509/cert_provider_spec.rb:
```diff
@@ -0,0 +1,305 @@
+        expect(File.read(path)).to match(/\A-----BEGIN CERTIFICATE REQUEST-----.*?-----END CERTIFICATE REQUEST-----\Z/m)
+      end
+
+      it 'sets mode to 644'
```
unfinished test?

melissa commented on lib/puppet/x509/cert_provider.rb:
```diff
@@ -0,0 +1,186 @@
+  # @api private
+  def load_private_key_from_pem(pem)
+    # set a non-nil passphrase to ensure openssl doesn't prompt
+    OpenSSL::PKey::RSA.new(pem, '')
```
what happens if we do have an encrypted key? Is the user just out of luck?

joshcooper replied:
Yep, and that's the current behavior of puppet (if you point it at an encrypted private key, but the `Puppet[:passfile]` file doesn't exist.

joshcooper replied:
I also filed https://tickets.puppetlabs.com/browse/PUP-9466 to readd support for encrypted private keys. Current support is dubious, and puppetserver doesn't support it (when agent and server on the same host).

melissa commented on tasks/generate_cert_fixtures.rake:
```diff
@@ -0,0 +1,43 @@
+desc "generate cert fixtures"
+task(:gen_cert_fixtures) do
```
What's the use case for this? Is this just in case someone wants to update the test fixtures? It'd be nice if you could add a little more detail about why you added this new task and under what scenarios someone might want to run it

joshcooper replied:
I added some comments to this file.

joshcooper commented:
I think this addresses all of the feedback so far. Everything was fairly cosmetic, except the `CertProvider` methods now raise `Puppet::Error` instead of letting the `PemStore` exceptions escape. For example, `PemStore` can raise `Errno::EACCES` due to permission denied, but that's very specific to a file `read` system call. This way we can change the `CertProvider` implementation in the future, such as loading private keys from a TPM without breaking consumers of the API.

melissa commented:
jruby failure
```
  1) Puppet::X509::PemStore deleting raises EACCES if the file is undeletable
     Failure/Error:
       expect {
         subject.delete_pem(path)
       }.to raise_error(Errno::EACCES, /Permission denied/)
       expected Errno::EACCES with message matching /Permission denied/ but nothing was raised
```

joshcooper commented:
@melissa I updated this to resolve the JRuby issue

melissa commented:
jenkins please test this

joshcooper commented on spec/unit/x509/cert_provider_spec.rb:
```diff
@@ -0,0 +1,391 @@
+      end
+
+      it 'raises when invalid input is inside BEGIN-END block' do
+        pending('jruby NPE bug') if Puppet::Util::Platform.jruby?
```
Filed as https://github.com/jruby/jruby/issues/5619

joshcooper commented:
- [ ] add jruby bug ID to pending test
- [ ] base64 encode RSA private keys so it doesn't generate false positive

joshcooper commented:
acceptance passed in https://jenkins-platform.delivery.puppetlabs.net/job/platform_puppet-agent-extra_puppet-agent-integration-suite_pr/878/
